src/lib/lessons.ts

```diff
@@ -49,38 +77,155 @@ async function loadLessonMetadataFromSanity(slug: string) {
+}
+
+async function loadLessonMetadataFromGraphQL(slug: string, token?: string) {
+  const graphQLClient = getGraphQLClient(token)
+
+  try {
+    const {lesson: lessonMetadataFromGraphQL} = await graphQLClient.request(
+      loadLessonGraphQLQuery,
+      {slug},
+    )
+
+    return lessonMetadataFromGraphQL
+  } catch (e) {
+    // Likely a 404 Not Found error
+    console.log('Error fetching from GraphQL: ', e)
+
+    return {}
+  }
 }
 
 export async function loadLesson(
   slug: string,
   token?: string,
 ): Promise<LessonResource> {
   token = token || getAccessTokenFromCookie()
-  const graphQLClient = getGraphQLClient(token)
 
-  const variables = {
-    slug: slug,
-  }
-
-  const {lesson} = await graphQLClient.request(
-    loadLessonGraphQLQuery,
-    variables,
+  /******************************************
+   * Primary Lesson Metadata GraphQL Request
+   * ****************************************/
+  const lessonMetadataFromGraphQL = await loadLessonMetadataFromGraphQL(
+    slug,
+    token,
   )
 
-  const lessonMetadataFromSanity = await loadLessonMetadataFromSanity(slug)
-
-  const lessonMetadata = {...lesson, ...lessonMetadataFromSanity}
-
+  /**********************************************
+   * Load comments from separate GraphQL Request
+   * ********************************************/
   // comments are user-generated content that must come from the egghead-rails
   // backend, so load those separately from the rest of lesson metadata.
   const comments = await loadLessonComments(slug, token)
 
+  /*************************************
+   * Sanity Request for Lesson Metadata
+   * ***********************************/
+  // this will be used to override values from graphql
+  const lessonMetadataFromSanity = await loadLessonMetadataFromSanity(slug)
+
+  /*************************************
+   * Merge All Lesson Metadata Together
+   * ***********************************/
+  // with preference for data coming from Sanity
+  const lessonMetadata = mergeLessonMetadata(
+    lessonMetadataFromGraphQL,
+    lessonMetadataFromSanity,
+  )
+
+  // if we aren't able to find Lesson metadata at either source, throw an
+  // error.
+  if (isEmpty(lessonMetadata.slug)) {
+    throw new Error(`Unable to lookup lesson metadata (slug: ${slug})`)
+  }
+
   return {...lessonMetadata, comments} as LessonResource
 }
 
+const mergeLessonMetadata = (
+  lessonMetadataFromGraphQL: LessonResource,
+  lessonMetadataFromSanity: LessonResource,
+): LessonResource => {
+  // we can merge most of it together as is, but there are a few nested pieces
+  // that need to be handled either manually.
+  //
+  // e.g. if tags haven't been set yet on Sanity, they will appear as an empty
+  // array. With a standard spread, they empty tags from Sanity would override
+  // an actual list of tags from graphql. We can instead handle this manually
+  // by checking for `_.some()` and falling back to graphql if there aren't
+  // any.
```

Comment:
@jbranchaud Have you seen lodash's _.mergeWith method? It may be what we are looking for. It does a deep merge but also lets you compare values using a _customizer_ function, which could be used here to check for empty arrays

Reply:
I should write some tests for this since it involves a bit of logic.

Reply:
Added tests in this commit https://github.com/eggheadio/egghead-next/pull/1154/commits/7da0426b31d2f042fda1d7422b5fd3c42a7a29b1